### PR TITLE
K.prod forced to run on CPU with Tensorflow

### PR DESCRIPTION
**Pull request by guicho271828**
On Tensorflow, backpropagation for `tf.reduce_prod` is a CPU-only operation, which severely increases the amount of CPU-GPU communication. It slows down one of my models by an order of magnitude.

See https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_grad.py#L124
and the related discussions at

* https://github.com/tensorflow/tensorflow/issues/3957
* https://github.com/tensorflow/tensorflow/pull/3351

This is a fix for circumventing this issue with repeatedly applying `tf.cumprod`, an operation supported by GPU.

### CODE REVIEW

**Comment by guicho271828** [comment]
I now feel I should submit this to tensorflow instead.

**Comment by guicho271828** [comment]
It looks like tf.cumprod has the same issue as tensorflow/tensorflow#3351 .
The gradient of tf.cumprod may return NaN when some partial gradient contains zero.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_grad.py#L1009
